### PR TITLE
#1 Moved sphinx documentation scripts

### LINKED ISSUE #1: Move Sphinx Documentation scripts from bucketfs-utils-python.
Move the files used for generating the Sphinx documentation over from [bucketfs-utils-python](https://github.com/exasol/bucketfs-utils-python/tree/feature/38_publish_release_docs_to_github_pages), and add possibility to specify which GitHub repository should be used.
Because of ongoing trouble with the deploy-to-github-pages.sh shell script, please rewrite it in python3

### PR DESCRIPTION
**Pull request by MarleneKress79789**
Fixed #1

### CODE REVIEW

File: exasol_sphinx_github_pages_generator/deploy_github_pages.py [hunk 1]
```diff
@@ -72,16 +65,17 @@ def checkout_target_branch_as_worktree(target_branch, worktree, push_origin):
...
-
+        os.chdir(currentworkdir)
 
-def build_and_copy_documentation(build_dir, worktree, source_branch, source_dir):
+def build_and_copy_documentation(build_dir, worktree, source_branch, source_dir, module_path):
     print("Build with sphinx")
-    run(["sphinx-build", "-M", "html", source_dir, build_dir, "-W"])
+    currentworkdir = os.getcwd()
+    print("currentworkdir :" + currentworkdir)
+    run(["sphinx-apidoc", "-T", "-d", "1", "-e", "-o", "api", module_path])
```
Comment (tomuben) [inline exasol_sphinx_github_pages_generator/deploy_github_pages.py:74]: Would be good to add some documentation what this command actually does.

File: exasol_sphinx_github_pages_generator/deploy_github_pages.py [hunk 2]
```diff
@@ -72,16 +65,17 @@ def checkout_target_branch_as_worktree(target_branch, worktree, push_origin):
...
+    run(["sphinx-build", "-b", "html", "-W", source_dir, build_dir]) # TODO own path for source dir?
     print("Generated HTML Output")
-    html_output_dir = build_dir + "/html/"
-    print(f"Using html_output_dir={html_output_dir}")
-    #run([ls -la "$HTML_OUTPUT_DIR"])
+    print(f"Using html_output_dir={build_dir}")
+    run(["ls", "-la", build_dir])
```
Comment (tomuben) [inline exasol_sphinx_github_pages_generator/deploy_github_pages.py:78]: I believe this is just for debugging.....can you remove it?

File: exasol_sphinx_github_pages_generator/deploy_github_pages.py [hunk 3]
```diff
@@ -90,43 +84,53 @@ def build_and_copy_documentation(build_dir, worktree, source_branch, source_dir)
...
+    print(f"Copying HTML output {build_dir} to the output directory {output_dir}")
+    for obj in os.listdir(build_dir):
+        shutil.move(build_dir + "/" + str(obj), output_dir)
+    # TODO test correctness of : find "build_dir" -mindepth 1 -maxdepth 1 -exec mv -t "$OUTPUT_DIR" -- {} +
     print(f"Content of output directory {output_dir}")
-
-    # touch "$WORKTREE/.nojekyll"
-    # run([ls -la "$OUTPUT_DIR"])
+    open(f"{worktree}/.nojekyll", "w").close()
+    run(["ls", "-la", output_dir])
```
Comment (tomuben) [inline exasol_sphinx_github_pages_generator/deploy_github_pages.py:93]: Is this also for debugging?

File: exasol_sphinx_github_pages_generator/deploy_github_pages.py (hunk 2 above)
Comment (MarleneKress79789) [reply exasol_sphinx_github_pages_generator/deploy_github_pages.py:78]: It was also present in the original script. I can remove it, but i would like to keep it until i am done with writing the test.

File: exasol_sphinx_github_pages_generator/parser.py [hunk 4]
```diff
@@ -16,4 +16,6 @@ def __init__(self):
         parser.add_argument('--source_branch', type=str,
                             default="main",
                             help='current branch')
-        self.args = parser.parse_args()
+        parser.add_argument('--source_dir', type=str, default="./doc/", help="The doc directory your documentation fildes adn conf.py reside in")
```
Comment (tomuben) [inline exasol_sphinx_github_pages_generator/parser.py:19]:
```suggestion
        parser.add_argument('--source_dir', type=str, default="./doc/", help="The doc directory your documentation files and conf.py reside in")
```

File: exasol_sphinx_github_pages_generator/parser.py [hunk 5]
```diff
@@ -16,4 +16,6 @@ def __init__(self):
...
+        parser.add_argument('--module_path', type=str, default="../module_name", help="The path to the module the docu is beeing generated for")
```
Comment (tomuben) [inline exasol_sphinx_github_pages_generator/parser.py:20]: Does the default value make sense?

File: githooks/install.sh [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+REPO_DIR=$(git rev-parse --show-toplevel)
+REPO_DIR="$(readlink -f "${REPO_DIR}")"
+GIT_DIR="$REPO_DIR/.git"
+GIT_DIR="$(readlink -f "${GIT_DIR}")"
+
+if [[ ! -d "$GIT_DIR" ]]; then
+  if [[ -d "$REPO_DIR/../.git" ]]; then
+    GIT_DIR="$REPO_DIR/../.git"
+    GITHOOKS_PATH="$GIT_DIR/modules/sphinx-github-pages-generator/hooks"
+  else
```
Comment (tomuben) [inline githooks/install.sh:16]: When can this happen? If we are in a submodule? Please add a comment

File: pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,39 @@
+[tool.poetry]
+name = "exasol-sphinx-github-pages-generator"
+version = "0.1.0"
+description = "Generates Sphinx GitHub pages for a given Git Repository"
+
+license = "MIT"
+
+authors = [
+    "Marlene Kreß <marlene.kress@exasol.com>"
```
Comment (tomuben) [inline pyproject.toml:9]: I think you can add Torsten as well, he wrote the original code.

**Review by tomuben — COMMENTED** [review]
Please add also the poetry lock file.
And check my comments

File: exasol_sphinx_github_pages_generator/deploy_github_pages.py (hunk 1 above)
Comment (MarleneKress79789) [reply exasol_sphinx_github_pages_generator/deploy_github_pages.py]: done

File: githooks/install.sh (hunk 6 above)
Comment (MarleneKress79789) [reply githooks/install.sh:16]: The githooks are also copies from the old repository. I have a different issue in witch i will go through them in detail.

File: exasol_sphinx_github_pages_generator/parser.py (hunk 5 above)
Comment (MarleneKress79789) [reply exasol_sphinx_github_pages_generator/parser.py:20]: At the moment no default value makes sense here. I set these to clarify the usage a little, but i can also take them out.

File: pyproject.toml (hunk 7 above)
Comment (MarleneKress79789) [reply pyproject.toml:9]: done

File: exasol_sphinx_github_pages_generator/deploy_github_pages.py [hunk 8]
```diff
@@ -90,43 +90,53 @@ def build_and_copy_documentation(build_dir, worktree, source_branch, source_dir)
...
+    run(["git", "status"])
+    changes_exists = run(["git", "diff-index", "--quiet", "HEAD", "--"], capture_output=True, text=True)
+    if 1 == changes_exists.returncode:
+        print(f"committing changes because changes_exist is {changes_exists.returncode}")
+        run(["git", "commit", "--no-verify", "-m", f"Update documentation from source branch {source_branch} with commit id {current_commit_id}"])
     if push_origin != "" and push_enabled == "push":
         print(f"Git push {push_origin} {target_branch}")
         run(["git", "push", push_origin, target_branch])
-    #popd
+    os.chdir(currentworkdir)
```
Comment (tomuben) [inline exasol_sphinx_github_pages_generator/deploy_github_pages.py:121]: 1. Do we need to push of there is no change (if we didn't commit)?
2. Maybe we should print a log if we there was no change (change_exists != 1)

File: exasol_sphinx_github_pages_generator/parser.py (hunk 5 above)
Comment (tomuben) [reply exasol_sphinx_github_pages_generator/parser.py:20]: Ok, maybe add a TODO to not forget to change it later

File: exasol_sphinx_github_pages_generator/deploy_github_pages.py (hunk 2 above)
Comment (tomuben) [reply exasol_sphinx_github_pages_generator/deploy_github_pages.py:84]: Maybe add a TODO to not forget about it

**Review by tomuben — COMMENTED** [review]
Please check my comments

File: exasol_sphinx_github_pages_generator/parser.py (hunk 5 above)
Comment (MarleneKress79789) [reply exasol_sphinx_github_pages_generator/parser.py:21]: done

File: exasol_sphinx_github_pages_generator/deploy_github_pages.py (hunk 8 above)
Comment (MarleneKress79789) [reply exasol_sphinx_github_pages_generator/deploy_github_pages.py]: done

File: exasol_sphinx_github_pages_generator/deploy_github_pages.py (hunk 2 above)
Comment (MarleneKress79789) [reply exasol_sphinx_github_pages_generator/deploy_github_pages.py:86]: done

File: exasol_sphinx_github_pages_generator/deploy_github_pages.py [hunk 9]
```diff
@@ -72,16 +67,23 @@ def checkout_target_branch_as_worktree(target_branch, worktree, push_origin):
...
+    currentworkdir = os.getcwd()
+    print("currentworkdir :" + currentworkdir)
+    # automatically generates Sphinx sources inside the "api" directory that document
+    # the package found in "module_path"
+    # -T: not table of contents
+    # -e: put documentation for ech module on own page
+    run(["sphinx-apidoc", "-T", "-e", "-o", "api", module_path])
+    # Builds the Sphinx documentation. Generates html files inside "build_dir" using "source_dir"
+    # -W: Turns warnings into errors
+    run(["sphinx-build", "-b", "html", "-W", source_dir, build_dir])
```
Comment (tomuben) [inline exasol_sphinx_github_pages_generator/deploy_github_pages.py:83]: I believe you don't check the return-code of any of the run-methods. That will make error detection quiete hard if something goes wrong.
You could wrap the run-method in your own run-method which checks the returncode and raises an exception if the return code is not 0.

**Review by tomuben — APPROVED** [review]

**Pull request merged**